File: src/main/java/com/ballgame/GameObject.java
package com.ballgame;

import java.awt.Graphics2D;
import java.awt.Image;
import java.util.List;
import java.awt.image.*;
import java.awt.geom.AffineTransform; 

public abstract class GameObject {
    protected double x; 
    protected double y; 

    protected double imagePaddingFactor = 1.0d; 
    protected BufferedImage image; 
    protected double imageFrame;
    protected int imageFPS; 

    protected List<BufferedImage> sprites; 
    protected List<BufferedImage> transformedSprites; 

    public GameObject(int x, int y) {
        this.x = x; 
        this.y = y; 
    }

    public GameObject() {
        this(0, 0);
    }

    public double getImageFrame() { return this.imageFrame; }

    public List<BufferedImage> getSprites() { return this.sprites; }

    public double getY() { return this.y; }
 
    public double getX() { return this.x; }

    protected void initSprites(List<BufferedImage> sprites, int imageFPS, double imagePaddingFactor) {
        if (sprites.size() > 0) {
            this.transformedSprites = AssetManager.clone(BufferedImage.class, sprites); 
            this.imagePaddingFactor = imagePaddingFactor; 
            this.sprites = sprites;
            this.imageFPS = imageFPS; 
            this.imageFrame = 0;
            this.image = this.sprites.get((int) this.imageFrame);
        }
    }

    protected void initSprites(List<BufferedImage> sprites) {
        this.initSprites(sprites, 24, 1.0d);
    }

    protected boolean incrementImageFrame(double deltaTime, double previousTime) {
        int prevFrame = (int) this.imageFrame;
        if ((int) this.imageFrame < this.sprites.size() - 1) {
            this.imageFrame = this.imageFrame + (this.imageFPS * (deltaTime - previousTime));
        } 
        else {
            this.imageFrame = 0;
        }

        if (prevFrame == (int) this.imageFrame) {
            return false; 
        }
        this.image = this.sprites.get((int) this.imageFrame);
        return true;
    }

    public double getDistance(double x, double y, double otherX, double otherY) {
        double deltaX = otherX - x; 
        double deltaY = otherY - y; 
        return Math.sqrt(Math.pow(deltaX, 2) + Math.pow(deltaY, 2));
    }

    public BufferedImage rotateImage(Image image, double angle) {
        double rads = Math.toRadians(angle);
        double sin = Math.abs(Math.sin(rads)), cos = Math.abs(Math.cos(rads));
        int w = image.getWidth(null);
        int h = image.getHeight(null);
        int newWidth = (int) Math.floor(w * cos + h * sin * (1 / this.imagePaddingFactor));
        int newHeight = (int) Math.floor(h * cos + w * sin * (1 / this.imagePaddingFactor)) ;
    
        BufferedImage rotated = new BufferedImage(newWidth, newHeight, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g2d = rotated.createGraphics();
        AffineTransform at = new AffineTransform();
        at.translate((newWidth - w) / 2, (newHeight - h) / 2);
    
        int x = w / 2;
        int y = h / 2;
    
        at.rotate(rads, x, y);
        g2d.setTransform(at);
        g2d.drawImage((BufferedImage) image, 0, 0, null);
        g2d.dispose();

        return rotated;
    }

    public void setImageFrame(double frame) {
        this.imageFrame = frame; 
    }

    public void setTransform(double x, double y) {
        this.x = x; 
        this.y = y; 
    }

    public double setTransformX(double x) {
        this.x = x; 
        return this.x; 
    }

    public double setTransformY(double y) {
        this.y = y; 
        return this.y; 
    }
    
    public abstract void update(double deltaTime, double previousTime);

    public abstract void render(Graphics2D g2d);
}

File: src/main/kotlin/com/ballgame/GameObject.kt
package com.ballgame

import com.ballgame.AssetManager.Companion.clone
import java.awt.Graphics2D
import java.awt.Image
import java.awt.geom.AffineTransform
import java.awt.image.BufferedImage
import kotlin.math.*

abstract class GameObject @JvmOverloads constructor(x: Int = 0, y: Int = 0) {
	var x: Double
		protected set
	var y: Double
		protected set
	protected var imagePaddingFactor = 1.0
	protected var image: BufferedImage? = null
	var imageFrame = 0.0
	private var imageFPS = 0
	var sprites: List<BufferedImage>? = null
		protected set
	protected lateinit var transformedSprites: MutableList<BufferedImage>

	init {
		this.x = x.toDouble()
		this.y = y.toDouble()
	}

	protected fun initSprites(sprites: List<BufferedImage>, imageFPS: Int = 24, imagePaddingFactor: Double = 1.0) {
		if (sprites.isNotEmpty()) {
			transformedSprites = clone(BufferedImage::class.java, sprites)
			this.imagePaddingFactor = imagePaddingFactor
			this.sprites = sprites
			this.imageFPS = imageFPS
			imageFrame = 0.0
			image = this.sprites!![imageFrame.toInt()]
		}
	}

	protected fun incrementImageFrame(deltaTime: Double, previousTime: Double): Boolean {
		val prevFrame = imageFrame.toInt()
		imageFrame =
			(if (imageFrame.toInt() < (sprites!!.size - 1)) imageFrame + (imageFPS * (deltaTime - previousTime)) else 0.0)
		if (prevFrame == imageFrame.toInt()) return false
		image = sprites!![imageFrame.toInt()]
		return true
	}

	fun getDistance(x: Double, y: Double, otherX: Double, otherY: Double) =
		sqrt((otherX - x).pow(2) + (otherY - y).pow(2))

	fun rotateImage(image: Image, angle: Double): BufferedImage {
		val rads = Math.toRadians(angle)
		val sin = abs(sin(rads))
		val cos = abs(cos(rads))
		val w = image.getWidth(null)
		val h = image.getHeight(null)
		val newWidth = floor((w * cos) + (h * sin * (1 / imagePaddingFactor))).toInt()
		val newHeight = floor((h * cos) + (w * sin * (1 / imagePaddingFactor))).toInt()
		val rotated = BufferedImage(newWidth, newHeight, BufferedImage.TYPE_INT_ARGB)
		val g2d = rotated.createGraphics()
		val at = AffineTransform()
		at.translate(((newWidth - w) / 2).toDouble(), ((newHeight - h) / 2).toDouble())
		val x = w / 2
		val y = h / 2
		at.rotate(rads, x.toDouble(), y.toDouble())
		g2d.transform = at
		g2d.drawImage(image as BufferedImage, 0, 0, null)
		g2d.dispose()
		return rotated
	}

	fun setTransform(x: Double, y: Double) {
		this.x = x
		this.y = y
	}

	fun setTransformX(x: Double): Double {
		this.x = x
		return this.x
	}

	fun setTransformY(y: Double): Double {
		this.y = y
		return this.y
	}

	abstract fun update(deltaTime: Double, previousTime: Double)
	abstract fun render(g2d: Graphics2D?)
}
